components/camel-aws/camel-aws2-kinesis/src/main/java/org/apache/camel/component/aws2/kinesis/Kinesis2Producer.java

```diff
@@ -47,11 +53,28 @@ public Kinesis2Endpoint getEndpoint() {
 
     @Override
     public void process(Exchange exchange) throws Exception {
+        Exception exceptionThrownByPutRecord = this.exceptionThrownByPutRecord;
+        if (exceptionThrownByPutRecord != null) {
+            this.exceptionThrownByPutRecord = null;
+            throw exceptionThrownByPutRecord;
+        }
+
         PutRecordRequest request = createRequest(exchange);
-        PutRecordResponse putRecordResult = connection.getClient(getEndpoint()).putRecord(request);
-        Message message = getMessageForResponse(exchange);
-        message.setHeader(Kinesis2Constants.SEQUENCE_NUMBER, putRecordResult.sequenceNumber());
-        message.setHeader(Kinesis2Constants.SHARD_ID, putRecordResult.shardId());
+        if (getEndpoint().getConfiguration().isAsyncClient()) {
```

Comment:
This is wrong, as you cannot use async client in a camel producer that is fully synchronous. Camel needs to be aware that this is async and route accordingly - this kind of work is complex and takes time to understand how to do it.

Reply:
Thanks for the comments. Keep using synchronous client.